src/ncp/ncp_spinel.cpp

```diff
@@ -102,6 +111,130 @@ otbrError NcpSpinel::SpinelDataUnpack(const uint8_t *aDataIn, spinel_size_t aDat
     return error;
 }
 
+void NcpSpinel::DatasetSetActiveTlvs(const otOperationalDatasetTlvs &aActiveOpDatasetTlvs,
+                                     AsyncResultReceiver             aReceiver)
+{
+    otError              error = OT_ERROR_NONE;
+    otOperationalDataset dataset;
+    otIp6Address         addrMlPrefix;
+    uint8_t              flagsSecurityPolicy[2];
+
+    VerifyOrExit(mDatasetSetActiveResultReceiver == nullptr);
+
+    SuccessOrExit(error = otDatasetParseTlvs(&aActiveOpDatasetTlvs, &dataset));
+    GetFlagsFromSecurityPolicy(&dataset.mSecurityPolicy, flagsSecurityPolicy, sizeof(flagsSecurityPolicy));
+    memcpy(addrMlPrefix.mFields.m8, dataset.mMeshLocalPrefix.m8, 8);
+    memset(addrMlPrefix.mFields.m8 + 8, 0, 8);
+
+    error = SetProperty(
+        SPINEL_PROP_THREAD_ACTIVE_DATASET,
+        SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT_PACKED_S SPINEL_DATATYPE_UINT64_S)
+            SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT_PACKED_S SPINEL_DATATYPE_DATA_S)
+                SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT_PACKED_S SPINEL_DATATYPE_UTF8_S)
+                    SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT_PACKED_S SPINEL_DATATYPE_DATA_S)
+                        SPINEL_DATATYPE_STRUCT_S(
+                            SPINEL_DATATYPE_UINT_PACKED_S SPINEL_DATATYPE_IPv6ADDR_S SPINEL_DATATYPE_UINT8_S)
+                            SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT_PACKED_S SPINEL_DATATYPE_UINT16_S)
+                                SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT_PACKED_S SPINEL_DATATYPE_UINT8_S)
+                                    SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT_PACKED_S SPINEL_DATATYPE_DATA_S)
+                                        SPINEL_DATATYPE_STRUCT_S(SPINEL_DATATYPE_UINT_PACKED_S SPINEL_DATATYPE_UINT16_S
+                                                                     SPINEL_DATATYPE_UINT8_S SPINEL_DATATYPE_UINT8_S),
+        SPINEL_PROP_DATASET_ACTIVE_TIMESTAMP, dataset.mActiveTimestamp.mSeconds, SPINEL_PROP_NET_NETWORK_KEY,
+        dataset.mNetworkKey.m8, sizeof(dataset.mNetworkKey.m8), SPINEL_PROP_NET_NETWORK_NAME, dataset.mNetworkName.m8,
+        SPINEL_PROP_NET_XPANID, dataset.mExtendedPanId.m8, sizeof(dataset.mExtendedPanId.m8),
+        SPINEL_PROP_IPV6_ML_PREFIX, &addrMlPrefix, OT_IP6_PREFIX_BITSIZE, SPINEL_PROP_MAC_15_4_PANID, dataset.mPanId,
+        SPINEL_PROP_PHY_CHAN, dataset.mChannel, SPINEL_PROP_NET_PSKC, dataset.mPskc.m8, sizeof(dataset.mPskc.m8),
+        SPINEL_PROP_DATASET_SECURITY_POLICY, dataset.mSecurityPolicy.mRotationTime, flagsSecurityPolicy[0],
+        flagsSecurityPolicy[1]);
```

Comment:
- It's kind of counter-intuitive that the items at the same level have different indentions. Not sure if this is a defect of the clang-format when dealing with macros.
- There might be too many repetitions of `SPINEL_DATATYPE` and `SPINEL_PROP` which makes it harder to notice what the actual data is.

Reply:
Yeah this is affected by clang-format... I don't find any good ways to handle this.

Reply:
In the OpenThread NCP code, we use the `Spinel::Decoder` class, which offers convenient helper methods for decoding and reading different data types (`Read{Type}()`) from the received Spinel frames.

- This approach simplifies the code and enhances readability.
- It also helps prevent potential errors where the input variables passed might not align with the expected `SPINEL_DATATYPE` specified in the format string (e.g., passing a `uint32_t` with `SPINEL_DATA_TYPE_UINT16`).

We could consider adopting the same approach here as well. We also have the `Spinel::Encoder` class for constructing spinel frames.

Reply:
I have updated to to `Spinel::Encoder`. I think it's much nicer now. Please take a look at the usage of `Spinel::Encoder`.

Reply:
Thank you! Looks great 😃